com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/eval/JdtEvaluationProvider.java

```diff
@@ -68,68 +69,53 @@ public void initialize(IDebugAdapterContext context, Map<String, Object> props)
     }
 
     @Override
-    public CompletableFuture<Value> evaluate(String expression, ThreadReference thread, int depth) {
-        CompletableFuture<Value> completableFuture = new CompletableFuture<>();
-        String projectName = (String) options.get(Constants.PROJECTNAME);
-        if (debugTarget == null) {
-            if (project == null) {
-                if (StringUtils.isBlank(projectName)) {
-                    logger.severe("Cannot evaluate when project is not specified.");
-                    completableFuture.completeExceptionally(new IllegalStateException("Please specify projectName in launch.json."));
-                    return completableFuture;
-                }
-                project = JdtUtils.getJavaProject(projectName);
-            }
-
-            if (project == null) {
-                completableFuture.completeExceptionally(new IllegalStateException(String.format("Project %s cannot be found.", projectName)));
-                return completableFuture;
-            }
-            if (launch == null) {
-                launch = createILaunchMock(project);
-            }
+    public CompletableFuture<Value> evaluateForBreakpoint(IBreakpoint breakpoint, ThreadReference thread, Map<IBreakpoint, Object> breakpointExpressionMap) {
```

Comment:
What's the difference of evaluating and evaluating for breakpoints?

Reply:
Because conditional breakpoint will be executed multiple time inside a loop(eval and cont), and the condition and location don't change, we need to optimize the conditional breakpoint, consider the following case, the i == 100 will be evaluated 999 times before the breakpoint is to hit.

 for (; i < 1000 + 1; i++) {
            if (i == 1000)  { # conditional BP of i == 1000
                System.out.println( "1111" );        
            } 
        }